Fix division by zero in MrDMD class

rfmiotto opened the pull request:
Dear,

As described in issue #122, here follows a proposal for fixing the division by zero in the multi-resolution DMD. The problem was being caused by the null eigenpairs in the `fit` method when the number of resolution levels defined by the user was large.

In the proposed pull request, a refactor was made only inside `fit` method. It is worth mentioning that this fix does not alter the solutions of the method, but avoids the zero-valued modes that were causing the problem in the reconstruction.

This pull request also includes a solution for the eventual `MemoryError` that raises when dealing with large matrices and large number of resolution levels. 

Best,
Renato Miotto

mtezzele commented:
Hi @rfmiotto! Thank you for your contribution. I kindly ask you to provide a couple of unit tests that check the while and the try except you added. Moreover at the end please squash the commits.

rfmiotto commented:
Dear @mtezzele, I squashed the commits and added a test for the maximum level, as requested.

I don't know if there is an efficient way to test for the `MemoryError` exception without creating a huge snapshot matrix. This approach can be very time consuming for the CI checks, I guess, so I dropped the idea.

**Note:** As you can see, some of the <ins>tests were changed</ins>. That was necessary in order to maintain the variables consistent. The code mimics the recursion process by adding and popping elements (the bins) into a queue during the while loop and <ins>not</ins> the level as stated before by the `current_level` variable. A new definition of the levels and `max_level` was adopted in this commit, and consequently, it was necessary to make some changes in the code, including the unit tests. The figure below tries to clarify how the levels are being defined now:

![IMG_20200304_244930385](https://user-images.githubusercontent.com/35971460/75850365-22a62900-5db5-11ea-8261-30a50ec52278.jpg)
In this figure, the purple numbers inside the tree structure indicate the index of each bin. These are the values returned by the `_index_list` method. 

The resolution level starts at zero. Hence, when one sets `max_level = 3`, he is defining the levels to be [0, 1, 2, **3**], which contains 4 items.

Please, let me know if you need anything else, or if you have any questions about this solution.
Best regards
